File: examples/python/receive_payload.py
import asyncio
import os
import json
from hermes import Config, HermesClient

ASCII_ART = r"""
    __  ___                 _                 __  __                             
   / / / (_)___  ____  ____(_)__  _______    / / / /__  _________ ___  ___  _____
  / /_/ / / __ \/ __ \/ __ \ / / / / ___/   / /_/ / _ \/ ___/ __ `__ \/ _ \/ ___/
 / __  / / /_/ / /_/ / /_/ / / /_/ (__  )  / __  /  __/ /  / / / / / /  __(__  ) 
/_/ /_/_/ .___/ .___/ .___/_/\__,_/____/  /_/ /_/\___/_/  /_/ /_/ /_/\___/____/  
       /_/   /_/   /_/                                                           

    [ Hippius Hermes - P2P Mass Data Messenger ]
"""

def on_message_received(action: str, sender_ss58: str, payload_bytes: bytes):
    """Callback fired directly from the Rust engine's multi-threaded QUIC sockets into Python."""
    print(f"\n[>>>] INCOMING SECURE MESSAGE FROM {sender_ss58}")
    print(f"      Action: {action}")

    try:
        meta = json.loads(payload_bytes.decode('utf-8'))
        file_hash = meta.get('hash', 'UNKNOWN')
        encrypted_keys = meta.get('encrypted_keys', [])

        print(f"      HCFS Hash/URL: {file_hash}")
        print(f"      Cipher Key Payload: <{len(encrypted_keys)} bytes>")
        print(f"[*] Secure Control Message verified. If this is a direct P2P payload, check your local data dir!")

    except Exception as e:
        print(f"      Raw Payload: {payload_bytes}")

async def main():
    print(ASCII_ART)
    print("[*] Initializing Hermes Listener Daemon...")

    key_path = os.environ.get("NODE_SECRET_KEY_PATH", "/path/to/iroh.key")
    ss58_address = os.environ.get("SS58_ADDRESS", "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY")
    api_token = os.environ.get("HIPPIUS_API_TOKEN", "your_hippius_api_token")
    data_dir = os.path.join(os.getcwd(), ".hermes_data")

    config = Config(key_path, ss58_address, api_token, data_dir)

    client = await HermesClient.create(config)
    print("[+] Core Engine and Iroh Networking Online!")

    print("[*] Binding Iroh QUIC Router to local port. Awaiting incoming Substrate connections...")

    client.start_listener(on_message_received)

    try:
        while True:
            await asyncio.sleep(3600)
    except KeyboardInterrupt:
        print("\n[-] Shutting down Hermes QUIC listener.")

if __name__ == "__main__":
    asyncio.run(main())

File: examples/rust/src/bin/receive_payload.rs
use hippius_hermes_core::{Client, config::Config};
use std::path::PathBuf;

const ASCII_ART: &str = r#"
    __  ___                 _                 __  __                             
   / / / (_)___  ____  ____(_)__  _______    / / / /__  _________ ___  ___  _____
  / /_/ / / __ \/ __ \/ __ \ / / / / ___/   / /_/ / _ \/ ___/ __ `__ \/ _ \/ ___/
 / __  / / /_/ / /_/ / /_/ / / /_/ (__  )  / __  /  __/ /  / / / / / /  __(__  ) 
/_/ /_/_/ .___/ .___/ .___/_/\__,_/____/  /_/ /_/\___/_/  /_/ /_/ /_/\___/____/  
       /_/   /_/   /_/                                                           

    [ Hippius Hermes - P2P Mass Data Messenger ]
"#;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    println!("{}", ASCII_ART);
    println!("[*] Initializing Hermes Native Core Listener...");

    let config = Config {
        node_secret_key_path: "iroh.key".into(),
        ss58_address: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY".to_string(),
        api_token: "your_hippius_api_token".to_string(),
        storage_directory: PathBuf::from(".hermes_data"),
        rpc_url: "wss://rpc.hippius.network:443".to_string(),
        subnet_ids: vec![],
        enable_firewall: false,
        s3: None,
        pullweights_api_key: None,
        skip_identity_verification: false,
        enable_queue: false,
        encryption_key_path: None,
    };

    let client = Client::new(config).await?;
    println!("[+] Core Engine and Iroh Networking Online!");
    println!("[*] Binding Iroh QUIC Router to local port. Awaiting incoming Substrate connections...");

    // Hand over Rust callback closure directly into the rust tokio loop
    client.spawn_listener(
        |msg| {
            println!(
                "\n[>>>] INCOMING SECURE MESSAGE FROM {}",
                msg.sender_ss58
            );
            println!("      Action: {}", msg.action);

            if let Ok(meta) = serde_json::from_slice::<serde_json::Value>(&msg.payload) {
                let hash = meta["hash"].as_str().unwrap_or("UNKNOWN");
                let mut num_keys = 0;
                if let Some(arr) = meta["encrypted_keys"].as_array() {
                    num_keys = arr.len();
                }

                println!("      HCFS Hash: {}", hash);
                println!("      Cipher Key Payload: <{} bytes>", num_keys);
                println!("[*] To read the massive gradients, your worker would now cleanly download from https://arion.hippius.com/download/{}", hash);
                println!("[*] And decrypt it locally using your Wallet's Private Key!");
            } else {
                println!("      Raw Payload Bytes: {} bytes", msg.payload.len());
            }
        },
        None::<fn(String, String, String, u64)>,
    );

    // Keep the main process thread alive
    loop {
        tokio::time::sleep(tokio::time::Duration::from_secs(3600)).await;
    }
}
